### PR TITLE
Bagpiper-Integration-2: Trainer

### PR DESCRIPTION
**Pull request by jctian98**
SpeechLM currently replaces Hugging Face Qwen3 MoE experts with custom grouped-matmul and expert-parallel implementations. This change uses Transformers' native `grouped_mm` backend and removes SpeechLM's EP/DeepEP support.

- Configure native experts for both pretrained loading and pipeline-parallel replicas.
- Preserve the memory-efficient router loss and capture native router logits in PP, including with activation checkpointing.
- Delete `grouped_moe.py`, EP-specific sharding/prefetch/gradient-clipping branches, and their obsolete tests. Reject configurations that request EP.
- Keep unit coverage on CPU, organize tests by production source file, and preserve the native GPU dispatch comparison as opt-in coverage under `test_utils/speechlm/`.

The installation update is tracked separately in #6645. Validation used its dependencies in a fresh Python 3.12 conda environment: PyTorch 2.11.0+cu128, Transformers 5.14.1, and TorchTitan 0.2.2.

Validation:

- 122 related CPU tests passed with CUDA hidden and Hugging Face offline, including native-expert identity/configuration, router-loss gradients, PP activation checkpointing, and trainer behavior.
- The opt-in H100 BF16 test verifies two native grouped-MM calls and compares outputs, input gradients, and parameter gradients against eager experts. It passed with `torch.nn.functional.grouped_mm` and with that API temporarily hidden to exercise the `torch._grouped_mm` dispatch used by older PyTorch. PyTorch 2.9 itself was not installed for this check.
- Existing Transformers mocks are isolated from lazy-module replacement, so model and trainer tests pass together.
- Black, isort, and `git diff --check` passed.

Run the optional GPU check explicitly from the repository root:

```bash
python -m pytest -o addopts= test_utils/speechlm/test_native_moe_cuda.py
```

The diff includes deletion of the custom MoE implementation and movement of existing tests into source-specific files. Full Bagpiper training has not been run; the existing SpeechLM fused-loss return-value mismatch with Liger 0.8.2 is outside this MoE/EP change.

### CODE REVIEW

**Review by sw005320 — APPROVED** [review]
LGTM!
This is the right decision.

**Comment by sw005320** [comment]
@jctian98, can you deal with [coderabbitai](https://github.com/apps/coderabbitai)'s reviews?

**Comment by sw005320** [comment]
Thanks! Move to the next!
